rfc/0001-policy-server-custom-resource.md

```diff
@@ -0,0 +1,415 @@
+
+
+## Welcome back, cert-manager?
+
+As discussed before, we must have a mutation webhook to
+process `PolicyServer` and `ClusterAdmissionPolicy`
+resources.
+
+This is extremely important, because we must ensure we have
+all the finalizers set and the user is not allowed harm himself
+(eg: rename a PolicyServer, delete the `default` PolicyServer,...).
+
+The controller itself can act as a mutation webhook, that's a common
+practice with Kubebuilder.
+
+However, the webhook endpoint must be secured with a TLS certificate, which
+has to be somehow provided.
+
+This is usually done by leveraging [cert-manager](https://cert-manager.io/),
+which must be previously installed on the cluster.
+
+Should we just bite the bullet and introduce this external dependency? After all,
+cert-manager is already mandatory to install many other Kubernetes add-ons.
+
+Maybe we can extend our helm chart to have these configuration knobs:
+
+  * Install cert-manager
+  * Do not install cert-manager, because it already exists. I think this
+    should be the recommended option for production deployments
+  * Generate the secrets via helm. This [seems doable](https://medium.com/nuvo-group-tech/move-your-certs-to-helm-4f5f61338aca).
+    We would not be able to rotate them, this is definitely something for quick evaluation
+    of Kubewarden. That's not meant to be used in production
+
+However, if we introduce the cert-manager dependency, we could also leverage that to
+create the certificates used by the Policy Server instances.
+
+
+## Kubewarden controller developer experience
+
+Right now, to develop the controller, we just run `make run` and that's it.
+
+If we bring the webhook endpoint back, the local development might become more cumbersome.
+This needs more research, but it's not a blocking point IMHO.
+
+# Alternatives
+[alternatives]: #alternatives
+
+> - What other designs/options have been considered?
+> - What is the impact of not doing this?
+
+A simple approach could be to stick with the current architecture, but
+perform these changes:
+
+  * Controller: do not create the Policy Server Deployment anymore
+  * Helm chart: take care of creating the Policy Server Deployment
+
+By doing that we will have one single place where we can maintain all the Policy Server
+configurations: the helm chart. Doing changes to the helm chart is definitely faster
+than writing code inside of the controller.
+
+This will potentially solve the 1st problem we have: provide a nice UX to configure
+Policy Server, plus it would reduce the maintenance efforts on our side.
+
+This wouldn't however solve the scalability and resiliency problems of Policy Server.
+
+Last but not least, from past experience (Cloud Foundry work at SUSE, other community
+projects), moving things on helm is tempting (you can go faster), but over a certain
+limit this will not scale. It will become hard to maintain the helm chart and impossible
+to add all the business logic you might need.
```

Comment:
Also, there's the point of updating the policy list today by changing the configuration is not possible and can be problematic if done wrong. With the new proposal, this UX is in a safe spot now.

Reply:
Sorry, I don't get it. What do you mean with "policy list", are you referring to the contents of the `policy.yml` file that we feed to policy-server?

If that the case, this list is not meant to be managed by the user at the "raw" level (aka, changing the contents of the ConfigMap). The user manages that by using the ClusterAdmissionPolicy resource.

Am I missing something? :thinking:

Reply:
> Sorry, I don't get it. What do you mean with "policy list", are you referring to the contents of the policy.yml file that we feed to policy-server?

Yes.

We are enumerating the advantages of the new implementation in comparison to the old one. Take this from this RFC:

> Currently, user-made changes to the policy-server ConfigMap (e.g. manual edit via kubectl, or helm upgrade) do not lead to the Policy Server Deployment to be updated. Moreover, when changing the policy-server ConfigMap, the user must be really careful to not alter the contents of the policies.yml key. That can result in catastrophic outcome (all policies being dropped by the Policy Server, leading to massive failures to all the admission reviews, leading to potential denial of service against the Kubernetes cluster).

This is also something we can mention briefly here again I think.

Reply:
Addressed